contracts/reward/RewardsDistributor.sol

```diff
@@ -0,0 +1,176 @@
+   * during merkle tree generation
+   * @param user wallet address of reward beneficiary
+   * @param tokens array of tokens claimable by reward beneficiary
+   * @param cumulativeAmounts cumulative token amounts claimable by reward beneficiary
+   * @param merkleProof merkle proof of claim
+   * @return true if valid claim, false otherwise
+   **/
+  function isValidClaim(
+    uint256 cycle,
+    uint256 index,
+    address user,
+    IERC20Ext[] calldata tokens,
+    uint256[] calldata cumulativeAmounts,
+    bytes32[] calldata merkleProof
+  ) external view override returns (bool) {
+    if (cycle != merkleData.cycle) return false;
+    bytes32 node = keccak256(abi.encodePacked(cycle, index, user, tokens, cumulativeAmounts));
+    return MerkleProof.verify(merkleProof, merkleData.root, node);
+  }
+
+  /** 
+   * @dev Claim accumulated rewards for a set of tokens at a given cycle number
+   * @param cycle cycle number
+   * @param index user reward info index in the array of reward info
+   * during merkle tree generation
+   * @param user wallet address of reward beneficiary
+   * @param tokens array of tokens claimable by reward beneficiary
+   * @param cumulativeAmounts cumulative token amounts claimable by reward beneficiary
+   * @param merkleProof merkle proof of claim
+   * @return claimAmounts actual claimed token amounts sent to the reward beneficiary
+   **/
+  function claim(
+    uint256 cycle,
+    uint256 index,
+    address user,
+    IERC20Ext[] calldata tokens,
+    uint256[] calldata cumulativeAmounts,
+    bytes32[] calldata merkleProof
+  ) external override nonReentrant returns (uint256[] memory claimAmounts) {
+    require(cycle == merkleData.cycle, 'incorrect cycle');
+
+    // verify the merkle proof
+    bytes32 node = keccak256(abi.encodePacked(cycle, index, user, tokens, cumulativeAmounts));
+    require(MerkleProof.verify(merkleProof, merkleData.root, node), 'invalid proof');
+
+    claimAmounts = new uint256[](tokens.length);
+
+    // claim each token
+    for (uint256 i = 0; i < tokens.length; i++) {
+      // if none claimable, skip
+      if (cumulativeAmounts[i] == 0) continue;
+
+      uint256 claimable = cumulativeAmounts[i].sub(claimedAmounts[user][tokens[i]]);
+      if (claimable == 0) continue;
+
+      claimedAmounts[user][tokens[i]] = cumulativeAmounts[i];
+      claimAmounts[i] = claimable;
+      if (tokens[i] == ETH_TOKEN_ADDRESS) {
+        (bool success, ) = user.call{value: claimable}('');
+        require(success, 'eth transfer failed');
+      } else {
+        tokens[i].safeTransfer(user, claimable);
+      }
+    }
+    emit Claimed(cycle, user, tokens, claimAmounts);
+  }
+
+  /// @notice Propose a new root and content hash, only by admin
+  function proposeRoot(
+    uint256 cycle,
+    bytes32 root,
+    string calldata contentHash
+  ) external onlyAdmin {
+    require(cycle == merkleData.cycle.add(1), 'incorrect cycle');
```

Comment:
why not just `cycle = merkleData.cycle + 1` and remove the func param

Reply:
What are the odds the daoOperator accidentally sends multiple txns for the same merkle root? This acts as a safeguard. 

No strong opinion about it

Reply:
well, it's not likely to happen with many owners in the multisig, but ok we can add it as safeguard here